cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UsageHandler.java

```diff
@@ -111,6 +120,111 @@ public void getDatasetAppUsage(HttpRequest request, HttpResponder responder,
                                  @PathParam("dataset-id") String datasetId) {
     final DatasetId id = new DatasetId(namespaceId, datasetId);
     Set<? extends EntityId> ids = registry.getPrograms(id);
-    responder.sendJson(HttpResponseStatus.OK, ids);
+    responder.sendJson(HttpResponseStatus.OK,
+                       Collections2.transform(ids, BackwardCompatibility.MAKE_COMPATIBLE_WITH_35));
+  }
+
+  /**
+   * Class to wrap stuff needed to make {@link UsageHandler} backward compatible. See CDAP-7316
+   * Also, see {@link IdCompatibleEntityIds} documentation for more details.
+   *
+   * Note: This backward compatibility transformation does not have its separate unit tests since since the unit test
+   * for {@link UsageHandler} will deserialize the JSON for the below classes to {@link Id} class and any
+   * incompatibility will fail the JSON deserialization there.
+   */
+  private static final class BackwardCompatibility {
+
+    /**
+     * Function which transforms and {@link EntityId} {@link IdCompatibleEntityIds} for backward compatibility with
+     * {@link Id}. This only supports {@link EntityId} which can be returned by {@link UsageHandler}.
+     */
+    private static final Function<EntityId, IdCompatibleEntityIds> MAKE_COMPATIBLE_WITH_35 =
+      new Function<EntityId, IdCompatibleEntityIds>() {
+        @Override
+        public IdCompatibleEntityIds apply(final EntityId entity) {
+          if (entity instanceof DatasetId) {
+            return new IdCompatibleEntityIds.IdDatasetInstance((DatasetId) entity);
+          } else if (entity instanceof StreamId) {
+            return new IdCompatibleEntityIds.IdStream((StreamId) entity);
+          } else if (entity instanceof ProgramId) {
+            return new IdCompatibleEntityIds.IdProgram((ProgramId) entity);
+          } else {
+            throw new IllegalArgumentException(String.format("Entity %s cannot be converted into backward " +
+                                                               "compatible Id.", entity));
+          }
+        }
+      };
+
+    /**
+     * Just a simple POJO to support backward compatibility for {@link UsageHandler} See: CDAP-7316
+     * This class allows {@link UsageHandler} to return results in deprecated {@link Id} serialized form.
+     */
+    private abstract static class IdCompatibleEntityIds {
```

Comment:
Aren't these classes the same thing? I think we should either use the Id class in both places or use these classes in both places. Otherwise what is the point?